### ISSUE TITLE
Create require-jsdoc rule

### ISSUE THREAD

**Issue by yfr** [eslint/eslint#1842]
I would like to have a `requireJsDoc` option for valid-jsdoc. Could be `false` by default but would be nice to enforce jsDoc at all and then of course the correct jsDoc.

**Comment by nzakas** [comment]
Can you please provide more details for what you are proposing? Is not clear from your description.

**Comment by yfr** [comment]
Sure!

I want to make sure that every function gets it's jsDoc block. Maybe this would make more sense in it's own rule.

**Comment by nzakas** [comment]
We need to figure out which functions because you may not, for example, want jadoc on an iife or on an anonymous function passed into another function. 

We can probably day function declarations should have jsdoc comments. What else?

**Comment by yfr** [comment]
Sounds legit. Thought there might be some issues i did not think about. 

Maybe we could just start with `FunctionDeclaration`, `FunctionExpression` and `ArrowFunctionExpression` like in `valid-jsdoc`. These could also be just options.

**Comment by nzakas** [comment]
I don't think those are granular enough. Consider:

```js

function foo() {
   // most likely want to require for this
}

var foo = function() {
    // do we want to require for this? It's a FunctionExpression
};

foo(function() {
    // this is a FunctionExpression, too. 
});

(function() {
    // this is a FunctionExpression, too.
}());

// these?
var foo = () => {};
foo(() => {});
```

**Comment by yfr** [comment]
Ok. I see your point. Is there a way to make more precise decisions? I personally would like the first two to have jsDoc required. 

Do you think it's worth building a plugin? Or is this just to special?

**Comment by nzakas** [comment]
We can be more specific, we just have to know what to look for.

I like the idea of a `require-jsdoc` rule as part of ESLint. Just need to figure out the correct options.

**Comment by grabbou** [comment]
@nzakas what about starting with all prototype methods and class methods in ES6? Once the feature is up & running, we can figure out other places where it should work. If that's something you'd like to see working, can submit the proposal.

**Comment by xjamundx** [comment]
I think any function declarations at the root of a file would be what I'd personally want to see, but I can see how this depends drastically on the project and language at hand.

We may also talk to @nrsedat about his plugin to see if he has any ideas:
https://github.com/nrsedat/eslint-plugin-require-jsdoc

**Comment by nzakas** [comment]
@grabbou we need to stay with something that easily lends itself to extension later, that's the tricky part.

@xjamundx I think the naming is the big problem to solve here. Some of these are pretty ambiguous

**Comment by andreas-marschke** [comment]
Hi! I have a PR to run fail eslint on undocumented named functions it's here: https://github.com/eslint/eslint/pull/2980

If you'd like to have this issue resolved feel free to comment on the PR.

**Comment by sebastienbarre** [comment]
Thanks for creating this rule. Was wondering if I'm missing something here. Shouldn't it be triggered for `FunctionExpression` as well? This, for example, doesn't trigger any error, even though it has no doc block:
```
var promiseRejectError = function() {
  return Promise.reject(Error('boom'));
};
```
The source code seems [to mention it] (https://github.com/eslint/eslint/blob/master/lib/rules/require-jsdoc.js#L62), but `FunctionExpression` is not an [accepted configuration option](https://github.com/eslint/eslint/blob/master/lib/rules/require-jsdoc.js#L81).

Thanks

**Comment by gyandeeps** [comment]
Yup we do not support `FunctionExpression` (ref docs).

**Comment by sebastienbarre** [comment]
@gyandeeps Thanks. Any particular rationale? I was wondering about the code in the rule:
```
        "FunctionExpression": function(node) {
            if (options.MethodDefinition) {
                checkClassMethodJsDoc(node);
            }
        },
```

**Comment by JustinLivi** [comment]
Are there any plans to support function expressions? It seems very useful when documenting modules. For example, right now there's no way to consider this a problem:

```javascript
module.exports.foo = function foo() {
    return 'bar';
}
```

**Comment by vobujs** [comment]
+1 on this

**Comment by platinumazure** [comment]
Please open a new issue if you want an enhancement to this rule. Thanks!

**Comment by vobujs** [comment]
here it is: https://github.com/eslint/eslint/issues/5867
